### PR TITLE
Allow for dynamic hosts in CORS origin config

### PR DESCRIPTION
**Pull request by grega**
## Status

- Closes https://github.com/RaspberryPiFoundation/editor-api/issues/405

## What's changed?

This extends the current origin-setting functionality (literal origins set via a comma-separated environment variable) and allows the env var string to also contain regular expressions. These allow the origins to match dynamic patterns (eg. `https://*.projects-ui.pages.dev`) and ensures that configuration still remains in the environment as we don't want to assume that anyone running this application will want the same allowed origins configured as us (other than for `localhost` for local and test environments only).

eg. instead of currently having to set the following in `ALLOWED_ORIGINS`:

```
https://foo.raspberrypi.org, https://bar.raspberrypi.org, https://foo.editor-standalone-eyq.pages.dev, https://bar.editor-standalone-eyq.pages.dev
```

The following can be set:

```
/https:\/\/(?:[a-z0-9-]+\.)?raspberrypi\.org$/, /https:\/\/.+\.editor-standalone-eyq\.pages\.dev$/
```

Associated updates in Terraform allow for the origins to match the spec in the [issue](https://github.com/RaspberryPiFoundation/editor-api/issues/405) (eg. the addition of wildcards / regex matching):

* https://github.com/RaspberryPiFoundation/terraform/pull/898

### CODE REVIEW

File: config/initializers/cors.rb [hunk 1]
```diff
@@ -1,24 +1,31 @@
...
 
 Rails.application.config.middleware.insert_before 0, Rack::Cors do
   allow do
-    origins origins_array
-    resource '*', headers: :any, methods: %i[get post patch put delete], expose: ['Link']
+    # localhost and test domain origins
+    if Rails.env.development?
+      origins(%r{https?://localhost([:0-9]*)$})
+    elsif Rails.env.test?
+      origins(%r{https?://localhost([:0-9]*)$}, %r{https?://www\.example\.com$})
```
Comment (conorriches) [inline config/initializers/cors.rb:14]: apologies if I'm not getting something here, do we want example.com in here?

**Review by conorriches — APPROVED** [review]
I've left a comment for something I'm just double checking on, but this looks great

File: config/initializers/cors.rb (hunk 1 above)
Comment (grega) [reply config/initializers/cors.rb:14]: Good question! It was something I initially copied over from the reference badge-service implementation (I assume it's used in a few specs there) but don't think it's strictly needed here at all...

I'll drop it as it's not really adding anything to keep it around.

**Review by sra405 — APPROVED** [review]
👍 lgtm, just worth checking the corp middleware in particular in staging before deploying to prod

**Pull request merged**